content/en/blog/_posts/2025-05-09-start-sidecar-first/index.md

```diff
@@ -0,0 +1,225 @@
+---
+layout: blog
+title: "Start Sidecar First: How To Avoid Snags"
+date: 2025-05-05
+draft: true
+slug: start-sidecar-first
+author: Agata Skorupka (The Scale Factory)
+---
+
+From the [previous blog post](/blog/2025/04/22/multi-container-pods-overview/) on Multicontainer Pods you know what their job is, what are the main architectural patterns, and how they are implemented in Kubernetes. The main thing I’ll cover in this article is how to ensure that your sidecar containers start before the main app. It’s more complicated than you might think!
+
+## A gentle refresher
+
+Let’s just remind that the [v1.29.0 release of Kubernetes](/blog/2023/12/13/kubernetes-v1-29-release/) added native support for [sidecar containers](/docs/concepts/workloads/pods/sidecar-containers/), which can now be defined within the `spec.initContainers` field, but with `restartPolicy: Always` as in the example below (note that it is a partial snippet of the full Kubernetes manifest):
+
+```yaml
+initContainers:
+  - name: logshipper
+    image: alpine:latest
+    restartPolicy: Always
+    command: ['sh', '-c', 'tail -F /opt/logs.txt']
+    volumeMounts:
+    - name: data
+        mountPath: /opt
+```
+
+What’s the specifics of defining sidecars with an `spec.initContainers` block, rather than as a regular multi-container pod? `spec.initContainers` are always started first (before the main application) and terminated last (after the main application). Furthermore, when used with [Jobs](/docs/concepts/workloads/controllers/job/), sidecar would still be alive and potentially restarted after Job is done, and they will not block pod completion (see also [Pod sidecar containers tutorial](/docs/tutorials/configuration/pod-sidecar-containers/)).
+
+## The problem
+
+Now you know that defining a sidecar with this native approach will always start it before the main application. From the kubelet source code, we know that this often means [being started almost in parallel](https://medium.com/@marko.luksa/delaying-application-start-until-sidecar-is-ready-2ec2d21a7b74), and this is not always what we want to achieve. What we are really interested in is whether we can delay the start of the main application until the sidecar is not just started, but fully running and ready to serve.
```

Comment:
> Now you know that defining a sidecar with this native approach will always start it before the main application.

Consider rewording, e.g. "Now that you know a sidecar container always starts before the main..."

> From the kubelet source code, we know that this often means [being started almost in parallel](https://medium.com/@marko.luksa/delaying-application-start-until-sidecar-is-ready-2ec2d21a7b74)

Why not reference the source code instead of this 3rd party blog post?

Reply:
I agree referencing source code looks better so I've changed that, but to be honest I wanted to give the author of that article a reference.